**В задании предлагается с помощью Pandas ответить на несколько вопросов по данным репозитория UCI [Adult](https://archive.ics.uci.edu/ml/datasets/Adult)**

Уникальные значения признаков (больше информации по ссылке выше):
- age: continuous.
- workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- fnlwgt: continuous.
- education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- education-num: continuous.
- marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- sex: Female, Male.
- capital-gain: continuous.
- capital-loss: continuous.
- hours-per-week: continuous.
- native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- salary: >50K,<=50K

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv("https://raw.githubusercontent.com/aksenov7/Kaggle_competition_group/master/adult.data.csv")
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [4]:
# def married(row):
#   return "Married" in row
data["married"] = data["marital-status"].apply(lambda row: "Married" in row)

In [5]:
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country salary  \
0              2174             0              40  United-States  <=50K   
1                 0             0              13  United-States  <=50K   
2                 0             0              40  United-States  <=50K   
3                 0             0              40  United-States  <=50K   
4                 0             0              40           Cuba  <=50K   
...             ...           ...             ...            ...    ...   
32556             0             0              38  United-States  <=50K   
32557             0             0              40  United-States   >50K   
32558             0             0              40  United-States  <=50K   
32559             0             0              20  United-States  <=50K   
32560         15024             0              40  United-States   >50K   

       married  
0        False  
1         True  
2        False  
3         True  
4         True  
...        ...  
32556     True  
32557     True  
32558    False  
32559    False  
32560     True  

[32561 rows x 16 columns]

**1. Сколько мужчин и женщин (признак *sex*) представлено в этом наборе данных?**

In [8]:
data['sex'].value_counts()

Male      21790
Female    10771
Name: sex, dtype: int64

В этом наборе данных было представлено 21790 мужчин и 10771 женщин.

**2. Каков средний возраст (признак *age*) женщин?**

In [10]:
sr_age_female = data[data["sex"] == "Female"]['age'].mean()
print("Средний возраст женщин в данном наборе:", sr_age_female)

Средний возраст женщин в данном наборе: 36.85823043357163


**3. Какова доля граждан Германии (признак *native-country*)?**

In [13]:
fraction_people = data.groupby("native-country").size() / len(data)
print("Доля граждан Германии", fraction_people["Germany"])

Доля граждан Германии 0.004207487485028101


**4-5. Каковы средние значения и среднеквадратичные отклонения возраста тех, кто получает более 50K в год (признак *salary*) и тех, кто получает менее 50K в год?**

In [14]:
print(data.groupby('salary')['age'].mean())
print(data.groupby('salary')['age'].std())

salary
<=50K    36.783738
>50K     44.249841
Name: age, dtype: float64
salary
<=50K    14.020088
>50K     10.519028
Name: age, dtype: float64


**6. Правда ли, что люди, которые получают больше 50k, имеют как минимум высшее образование? (признак *education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* или *Doctorate*)**

In [18]:
higher_education = ['Bachelors', 'Prof-school', 'Assoc-acdm', 'Assoc-voc', 'Masters', 'Doctorate']
true_or_false = data[data.education.isin(higher_education)].salary.value_counts()['<=50K'] == 0
print("Правда ли, что люди, которые получают больше 50k, имеют как минимум высшее образование -", true_or_false)

Правда ли, что люди, которые получают больше 50k, имеют как минимум высшее образование - False


**7. Выведите статистику возраста для каждой расы (признак *race*) и каждого пола. Используйте *groupby* и *describe*. Найдите таким образом максимальный возраст мужчин расы *Amer-Indian-Eskimo*.**

In [23]:
table = data.groupby(['race', 'sex']).describe()['age']
print(table)

                             count       mean        std   min   25%   50%  \
race               sex                                                       
Amer-Indian-Eskimo Female    119.0  37.117647  13.114991  17.0  27.0  36.0   
                   Male      192.0  37.208333  12.049563  17.0  28.0  35.0   
Asian-Pac-Islander Female    346.0  35.089595  12.300845  17.0  25.0  33.0   
                   Male      693.0  39.073593  12.883944  18.0  29.0  37.0   
Black              Female   1555.0  37.854019  12.637197  17.0  28.0  37.0   
                   Male     1569.0  37.682600  12.882612  17.0  27.0  36.0   
Other              Female    109.0  31.678899  11.631599  17.0  23.0  29.0   
                   Male      162.0  34.654321  11.355531  17.0  26.0  32.0   
White              Female   8642.0  36.811618  14.329093  17.0  25.0  35.0   
                   Male    19174.0  39.652498  13.436029  17.0  29.0  38.0   

                             75%   max  
race               sex

In [33]:
dataframe = data.groupby(["race", "sex"]).describe().loc['Amer-Indian-Eskimo'].loc["Male"]
print("Максимальный возраст мужчин расы Amer-Indian-Eskimo:", dataframe["age"]["max"], "года.")

Максимальный возраст мужчин расы Amer-Indian-Eskimo: 82.0 года.


**8. Среди кого больше доля зарабатывающих много (>50K): среди женатых или холостых мужчин (признак *marital-status*)? Женатыми считаем тех, у кого *marital-status* начинается с *Married* (Married-civ-spouse, Married-spouse-absent или Married-AF-spouse), остальных считаем холостыми.**

In [35]:
salary_man = data[data['sex'] == 'Male'].query('salary == ">50K"')
married_man = len(salary_man[salary_man["marital-status"]=="Married-civ-spouse"]) + len(salary_man[salary_man["marital-status"]=="Married-spouse-absent"]) + len(salary_man[salary_man["marital-status"]=="Married-AF-spouse"])
print("Доля зарабатывающих больше 50к больше среди женатых, чем холостых: ", married_man > len(salary_man) - married_man)


Доля зарабатывающих больше 50к больше среди женатых, чем холостых:  True


**9. Какое максимальное число часов человек работает в неделю (признак *hours-per-week*)? Сколько людей работают такое количество часов и каков среди них процент зарабатывающих много?**

In [37]:
max_hours_work_in_week = data['hours-per-week'].describe()['max']
people_work = data[data["hours-per-week"] == max_hours_work_in_week]
print("Максимальное число часов, которое человек работает в неделю:", max_hours_work_in_week)
print("Количество людей, работающих такое же количество часов:", people_work.shape[0])
print("Процент людей зарабатывающих много:", round((people_work[people_work['salary'] == '>50K'].shape[0] / people_work.shape[0]) * 100, 1), "%")

Максимальное число часов, которое человек работает в неделю: 99.0
Количество людей, работающих такое же количество часов: 85
Процент людей зарабатывающих много: 29.4 %


**10. Посчитайте среднее время работы (*hours-per-week*) зарабатывающих мало и много (*salary*) для каждой страны (*native-country*).**

In [34]:
print("Среднее время работы зарабатывающих мало и много для каждой страны")
print(data.groupby(['native-country', 'salary'])['hours-per-week'].mean())

Среднее время работы зарабатывающих мало и много для каждой страны
native-country  salary
?               <=50K     40.164760
                >50K      45.547945
Cambodia        <=50K     41.416667
                >50K      40.000000
Canada          <=50K     37.914634
                            ...    
United-States   >50K      45.505369
Vietnam         <=50K     37.193548
                >50K      39.200000
Yugoslavia      <=50K     41.600000
                >50K      49.500000
Name: hours-per-week, Length: 82, dtype: float64
